### PR TITLE
Add timestep_spacing to schedulers

### PR DESCRIPTION
**Pull request by pcuenca**

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 1]
```diff
@@ -302,7 +302,7 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
 
         self.num_inference_steps = num_inference_steps
 
-        # "leading" and "trailing" corresponds to annotation of Table 1. of https://arxiv.org/abs/2305.08891
+        # "leading" and "trailing" corresponds to annotation of Table 2. of https://arxiv.org/abs/2305.08891
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_ddim.py:305]: missing "trailing" from ddim, will add later

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 2]
```diff
@@ -134,6 +141,8 @@ def __init__(
         dynamic_thresholding_ratio: float = 0.995,
         clip_sample_range: float = 1.0,
         sample_max_value: float = 1.0,
+        timestep_spacing: str = "leading",
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_ddpm.py:144]: default is `leading` here

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 3]
```diff
@@ -228,11 +237,31 @@ def set_timesteps(
...
-
-            step_ratio = self.config.num_train_timesteps // self.num_inference_steps
-            timesteps = (np.arange(0, num_inference_steps) * step_ratio).round()[::-1].copy().astype(np.int64)
             self.custom_timesteps = False
 
+            # "linspace", "leading", "trailing" corresponds to annotation of Table 2. of https://arxiv.org/abs/2305.08891
+            if self.config.timestep_spacing == "linspace":
+                timesteps = np.linspace(0, self.config.num_train_timesteps - 1, num_inference_steps).round()[
+                    ::-1
+                ].copy().astype(np.int64)
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_ddpm.py:246]: Need to be careful here as some schedulers use `float` as others use ints.

File: src/diffusers/schedulers/scheduling_lms_discrete.py [hunk 4]
```diff
@@ -117,6 +124,8 @@ def __init__(
         trained_betas: Optional[Union[np.ndarray, List[float]]] = None,
         use_karras_sigmas: Optional[bool] = False,
         prediction_type: str = "epsilon",
+        timestep_spacing: str = "linspace",
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_lms_discrete.py:127]: This one defaults to `linspace`

File: src/diffusers/schedulers/scheduling_lms_discrete.py [hunk 5]
```diff
@@ -140,16 +149,21 @@ def __init__(
...
         self.num_inference_steps = None
         self.use_karras_sigmas = use_karras_sigmas
         self.set_timesteps(num_train_timesteps, None)
         self.derivatives = []
         self.is_scale_input_called = False
 
+    @property
+    def init_noise_sigma(self):
+        # standard deviation of the initial noise distribution
+        if self.config.timestep_spacing == "linspace":
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_lms_discrete.py:162]: I think this should be the case for `trailing` too.

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 3 above)
Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_ddpm.py:246]: How about performing the type-casting after all the conditions have been traversed?

**Comment by patrickvonplaten** [comment]
> One problem is that `PNDMScheduler` defaults to `leading`, but others default to `linspace`. So for example when changing an scheduler after loading the pipeline, this would behave differently as before:
> 
> ```python
> import torch
> from diffusers import DiffusionPipeline
> from diffusers import HeunDiscreteScheduler
> 
> pipe = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
> pipe = pipe.to("cuda")
> 
> pipe.scheduler = HeunDiscreteScheduler.from_config(pipe.scheduler.config)
> # Here we are using heun with `leading` instead of the previous default `linspace`
> ```
> 
> @patrickvonplaten how do you think we should deal with this?

I thought about this too previously. Can we do the following here: We add a mechanism for values that are **not** loaded from the config: https://huggingface.co/runwayml/stable-diffusion-v1-5/blob/main/scheduler/scheduler_config.json and then default to the value in the `__init__` function that stores these as  a list in the `hidden_config_dict` here, under "_use_default_values: List[Any]" and then pop those values from the config when loading it again?

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 3 above)
Comment (pcuenca) [reply src/diffusers/schedulers/scheduling_ddpm.py:246]: Good idea, I'll check it out.

File: src/diffusers/configuration_utils.py [hunk 6]
```diff
@@ -503,6 +511,9 @@ def extract_init_dict(cls, config_dict, **kwargs):
         # 7. Define "hidden" config parameters that were saved for compatible classes
         hidden_config_dict = {k: v for k, v in original_dict.items() if k not in init_dict}
 
+        # 8. Take note of the parameters that were not present in the loaded config
+        hidden_config_dict["_use_default_values"] = used_default_keys
```
Comment (patrickvonplaten) [inline src/diffusers/configuration_utils.py:515]:
```suggestion
        hidden_config_dict["_use_default_values"] = set(init_dict) - expected_keys
```

File: src/diffusers/configuration_utils.py [hunk 7]
```diff
@@ -450,6 +454,10 @@ def extract_init_dict(cls, config_dict, **kwargs):
         else:
             compatible_classes = []
 
+        # Keys not present in the config - default values were used
+        # TODO: remove the ones passed in kwargs?
+        used_default_keys = set(expected_keys) - set(original_dict.keys())
+
```
Comment (patrickvonplaten) [inline src/diffusers/configuration_utils.py:460]:
```suggestion
```

Comment (patrickvonplaten) [reply src/diffusers/configuration_utils.py:460]: think we can remove this

File: tests/schedulers/test_schedulers.py [hunk 8]
```diff
@@ -202,6 +204,31 @@ def test_save_load_from_different_config_comp_schedulers(self):
         assert cap_logger_2.out == "{'f'} was not found in config. Values will be initialized to default values.\n"
         assert cap_logger_3.out == "{'f'} was not found in config. Values will be initialized to default values.\n"
 
+    def test_default_arguments_not_in_config(self):
+        pipe = DiffusionPipeline.from_pretrained(
+            "hf-internal-testing/tiny-stable-diffusion-pipe", torch_dtype=torch.float16
+        )
+        assert pipe.scheduler.__class__ == DDIMScheduler
+
+        # Default for PNDMScheduler
```
Comment (pcuenca) [inline tests/schedulers/test_schedulers.py:213]:
```suggestion
        # Default for DDIMScheduler
```

**Comment by patrickvonplaten** [comment]
PR looks great to me - however could we merge this into "main" and not the `sd_xl` branch so that in can go into this weeks' release?

**Comment by pcuenca** [comment]
Rebased in #3947.
